### PR TITLE
AdventureNBTSerializer reworked

### LINKED ISSUE #679: /give command throwing warning - SYSTEM_CHAT_MESSAGE
**Describe the bug**
When listening to the system chat message packet and modifying it, if I use the vanilla /give command, it throws a warning. This warning is not thrown in any other system messages that I have tried.

**Software brand**
Paper 1.20.4 - Latest

**Plugins**
Vault, PlaceholderAPI and my own plugin. (PacketEvents is shaded)

**How To Reproduce**
1. Listen to the system chat message packet.
2. Use the vanilla /give command.
3. [This](https://paste.helpch.at/osokeqiqud.rb) warning will pop up.

**Expected behavior**
Not to throw any warnings?

### PR DESCRIPTION
**Pull request by AbhigyaKrishna**
Tested on 1.20.4 with WrapperPlayServerSystemChatMessage

Fixes #679 and #673 

image

image

### CODE REVIEW

File: api/src/main/java/com/github/retrooper/packetevents/util/adventure/AdventureNBTSerializer.java [hunk 1]
```diff
@@ -0,0 +1,617 @@
...
+            // keybind
+            writer.writeUTF("keybind", ((KeybindComponent) component).keybind());
+        } else if (component instanceof NBTComponent<?, ?>) {
+            // nbt path
+            String nbtPath = ((NBTComponent<?, ?>) component).nbtPath();
+            writer.writeUTF("nbt", nbtPath);
+
+            // interpret
+            boolean interpret = ((NBTComponent<?, ?>) component).interpret();
+            if (interpret) { // defaults to true
```
Comment (booky10) [inline api/src/main/java/com/github/retrooper/packetevents/util/adventure/AdventureNBTSerializer.java:162]: This comment seems to be wrong (I think you just copied my wrong comment)

File: api/src/main/java/com/github/retrooper/packetevents/util/adventure/AdventureNBTSerializer.java [hunk 2]
```diff
@@ -0,0 +1,617 @@
...
+
+        NBTWriter writer = new NBTWriter(new NBTCompound());
+
+        Key font = style.font();
+        if (font != null) writer.writeUTF("font", font.asString());
+
+        TextColor color = style.color();
+        if (color != null) writer.writeUTF("color", serializeColor(color));
+
+        TextDecoration.State bold = style.decoration(TextDecoration.BOLD);
```
Comment (booky10) [inline api/src/main/java/com/github/retrooper/packetevents/util/adventure/AdventureNBTSerializer.java:266]: This whole TextDecoration block could now be replaced by just iterating all existing TextDecorations, as this no longer is one giant switch

File: api/src/main/java/com/github/retrooper/packetevents/util/adventure/AdventureNBTSerializer.java [hunk 3]
```diff
@@ -0,0 +1,617 @@
...
+
+        Style.Builder style = Style.style();
+        NBTReader reader = new NBTReader(input);
+
+        reader.useUTF("font", value -> style.font(Key.key(value)));
+        reader.useUTF("color", value -> {
+            TextColor color = deserializeColor(value);
+            if (color != null) style.color(color);
+        });
+        reader.useBoolean("bold", value -> style.decoration(TextDecoration.BOLD, TextDecoration.State.byBoolean(value)));
```
Comment (booky10) [inline api/src/main/java/com/github/retrooper/packetevents/util/adventure/AdventureNBTSerializer.java:211]: See comment below regarding serialization of TextDecoration

File: api/src/main/java/com/github/retrooper/packetevents/util/adventure/AdventureNBTSerializer.java [hunk 4]
```diff
@@ -0,0 +1,617 @@
+package com.github.retrooper.packetevents.util.adventure;
```
Comment (booky10) [inline api/src/main/java/com/github/retrooper/packetevents/util/adventure/AdventureNBTSerializer.java:1]: I think the license header is missing here

**Review by booky10 — COMMENTED** [review]
You could also add support for the "type" field, which vanilla now (sometimes?) uses: https://haste.pinofett.de/bqkxwnftqf.java

**Comment by AbhigyaKrishna** [comment]
> You could also add support for the "type" field, which vanilla now (sometimes?) uses: https://haste.pinofett.de/bqkxwnftqf.java

I don't think as of right now it is necessary to make the if cases complex.

File: api/src/main/java/com/github/retrooper/packetevents/util/adventure/AdventureNBTSerializer.java (hunk 2 above)
Comment (AbhigyaKrishna) [reply api/src/main/java/com/github/retrooper/packetevents/util/adventure/AdventureNBTSerializer.java]: Thanks for pointing those out

File: api/src/main/java/com/github/retrooper/packetevents/util/adventure/AdventureNBTSerializer.java [hunk 5]
```diff
@@ -1,3 +1,21 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
```
Comment (booky10) [inline api/src/main/java/com/github/retrooper/packetevents/util/adventure/AdventureNBTSerializer.java:18]: Are you sure this is ProtocolSupport? :smile:

Comment (AbhigyaKrishna) [reply api/src/main/java/com/github/retrooper/packetevents/util/adventure/AdventureNBTSerializer.java]: I copied it from NBT class file. My bad

**Review by booky10 — APPROVED** [review]
Looks good!

**Comment by AoElite** [comment]
@AbhigyaKrishna was the `org.intellij.lang.annotations.Subst` import intentional?

**Comment by AbhigyaKrishna** [comment]
> @AbhigyaKrishna was the `org.intellij.lang.annotations.Subst` import intentional?

Accidental import

**Pull request merged**